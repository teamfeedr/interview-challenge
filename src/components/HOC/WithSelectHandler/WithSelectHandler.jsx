import React from 'react';
import PropTypes from 'prop-types';

export const getSelectHandler = ({ state, updateState }) => ({ item }) => {
    const itemAlreadyInState = state.find((stateItem) => stateItem.id === item.id);

    if (itemAlreadyInState) {
        return;
    }

    updateState([...state, item]);
}

const WithSelectHandler = ({ selectedItemsState, updateSelectedItemsState, WrappedComponent, ...childProps }) => {
    const handleSelectItem = getSelectHandler({ state: selectedItemsState, updateState: updateSelectedItemsState });

    return (<WrappedComponent handleSelectItem={handleSelectItem} {...childProps} />);
}

WithSelectHandler.propTypes = {
    selectedItemsState: PropTypes.array.isRequired,
    updateSelectedItemsState: PropTypes.func.isRequired,
    WrappedComponent: PropTypes.func.isRequired
}

export default WithSelectHandler;