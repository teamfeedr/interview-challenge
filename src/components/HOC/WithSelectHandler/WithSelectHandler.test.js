import React from "react";
import { shallow } from "enzyme";

import WithSelectHandler, { getSelectHandler } from "./WithSelectHandler";

describe("the WithSelectHandler HOC", () => {
  it("renders a wrapped component with the handleItemSelect function prop", () => {
    const TestComponent = () => {
      <div />;
    };

    const rendered = shallow(
      <WithSelectHandler
        selectedItemsState={[]}
        updateSelectedItemsState={() => {}}
        WrappedComponent={TestComponent}
      />
    );

    expect(rendered).toMatchSnapshot();
  });
});

describe("the getSelectHandler function", () => {
  it("handles item selection", () => {
    const state = [
      {
        id: 1
      }
    ];

    const updateState = jest.fn();

    const handleItemSelect = getSelectHandler({ state, updateState });

    handleItemSelect({ item: { id: 1 } });

    expect(updateState).toHaveBeenCalledTimes(0);

    handleItemSelect({ item: { id: 3 } });

    expect(updateState).toHaveBeenCalledTimes(1);
    expect(updateState).toHaveBeenCalledWith([{ id: 1 }, { id: 3 }]);
  });
});
