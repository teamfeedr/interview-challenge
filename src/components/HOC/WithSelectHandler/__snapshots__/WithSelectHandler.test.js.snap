// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the WithSelectHandler HOC renders a wrapped component with the handleItemSelect function prop 1`] = `
<TestComponent
  handleSelectItem={[Function]}
/>
`;
